wagtail_newsletter/audiences.py

```diff
@@ -0,0 +1,134 @@
+from abc import abstractmethod
+
+from django.conf import settings
+from django.core.cache import caches
+from django.core.exceptions import ObjectDoesNotExist
+from queryish import Queryish, VirtualModel
+
+from . import campaign_backends
+
+
+class CachedApiQueryish(Queryish):
+    cache_prefix: str
+
+    @abstractmethod
+    def get_list(self):
+        raise NotImplementedError
+
+    def cache_key(self, pk):
+        return f"{self.cache_prefix}{pk}"
+
+    def get_instance(self, pk, **kwargs):
+        return self.model(id=pk, **kwargs)  # type: ignore
+
+    def parse_filters(self):
+        return dict(self.filters)
+
+    def get_detail(self, pk):
+        instance = self.get_list().get(pk, None)
+        if instance is None:
+            raise self.model.DoesNotExist  # type: ignore
+        return instance
+
+    def run_query(self):
+        cache = caches["default"]
+        filters = self.parse_filters()
+        if set(filters) == {"pk"}:
+            pk = filters["pk"]
+            cache_key = self.cache_key(pk)
+            kwargs = cache.get(cache_key)
+            if kwargs is None:
+                kwargs = self.get_detail(pk).to_json()
+                timeout = getattr(settings, "WAGTAIL_NEWSLETTER_CACHE_TIMEOUT", 300)
```

Comment:
I know this is a work in progress state, but I wonder if keeping track of the settings we introduce in a versioned text file couldn't be useful later when the time will have come to write some documentation for this package? :slightly_smiling_face:

Reply:
That's a good point, we'll certainly want to document all of those, so I've updated https://github.com/wagtail/wagtail-newsletter/issues/26 to mention this. I plan to grep the source code for `settings` and/or `WAGTAIL_NEWSLETTER_` and get a list of our settings that way.